File: src/liberary/account_lib/match_event/match_event.rs
use std::cmp::{PartialEq, Eq};
use serde::{Deserialize, Serialize};

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct MatchEvent {
    pub id: Option<i32>, // Not meant to be used in order to store in databank, solely for identification among shedule
    pub start_timestamp: u64,
    pub challenger_id: u64,
    pub opponent_id: u64,
    pub event_id: Option<u64>,
    pub status: MatchStatus,
    pub season: String
}

impl MatchEvent{

    pub fn get_id(&self) -> String {
        let mut lower_id = self.challenger_id.clone();
        let mut higher_id = self.opponent_id.clone();

        if lower_id > higher_id {
            lower_id = self.opponent_id.clone();
            higher_id = self.challenger_id.clone();
        }

        format!("{}V{}@{}", lower_id, higher_id, self.start_timestamp)
    }
}

#[derive(Clone, Serialize, Deserialize, Debug)]
pub enum MatchStatus {
    Requested,
    Confirmed,
    Finished,
    Declined
}

impl MatchStatus {
    pub fn from_status_code(code: i16) -> Result<MatchStatus, String> {
        match code {
            0 => Ok(MatchStatus::Requested),
            1 => Ok(MatchStatus::Confirmed),
            2 => Ok(MatchStatus::Finished),
            3 => Ok(MatchStatus::Declined),
            _ => Err("Invalid status code".to_string())
        }
    }
}

impl PartialEq for MatchEvent {
    fn eq(&self, other: &Self) -> bool {
        // Compare relevant fields for equality
        self.start_timestamp == other.start_timestamp
            && self.season == other.season
            && ((self.challenger_id == other.challenger_id && self.opponent_id == other.opponent_id)
                || (self.challenger_id == other.opponent_id && self.opponent_id == other.challenger_id))
    }
}

impl Eq for MatchEvent {} // Required for types implementing PartialEq

File: PORC-Front/src/models/match_event/MatchEvent.ts

import type { MatchEventRecv } from "./MatchEventRecv";

export interface MatchEvent {
    startDate: Date;
    endDate: Date;
    initiatorId: string;
    opponentId: string;
    status: MatchStatus;
    season: string;
}

export enum MatchStatus {
    Requested = 'Requested',
    Confirmed = 'Confirmed',
    Finished = 'Finished',
    Declined = 'Declined',
}

export enum ObservedMatchStatus {
    HasRequested = 'HasRequested', // hourglass_bottom
    HasBeenRequested = 'HasBeenRequested', // bell
    Confirmed = 'Confirmed', // calander_check
    Finished = 'Finished', // checkmark
    Declined = 'Declined', // calander_busy
    Unplaned = 'Unplaned',
    IsSelf = 'IsSelf', // calander
}

export function matchEventToRecv(val: MatchEvent): MatchEventRecv {
    return {
        id: null,
        start_timestamp: val.startDate.getTime() / 1000,
        challenger_id: val.initiatorId,
        opponent_id: val.opponentId,
        event_id: null,
        status: val.status,
        season: val.season
    } as MatchEventRecv
}

